Ninja: Add 2 new weapons

Xackii opened the pull request:
## About The Pull Request

Adds 2 new weapons for midround ninja:
https://youtu.be/FC2Ic6JuUDo
When spawned ninja have case that allow to pick 1/3 weapons: katana/glaive/kusarigama
1. Energy Glaive: https://youtu.be/4Rw6rwOiu7c
 Force 14
 Throwforce 40
 Block Chanse 70
 Teleport user when thrown. Make double attack
3. Energy Kusarigama: https://youtu.be/thYWTffet1g
  Kusarigama maden from 3 parts: head, chain and tail.
 Head(kama): 
   Force 12
   Throwforce 40
   Block chance 40 
 Tail(fundo):   
   Force 10
   Throwforce 20
   Block chance 60
   knock out weapon in melee attack 
   knockdown target in throw attack. 
 Chain(chain): Chain connected between head and tail. as soon as one of the friends moves far from the other, he in turn tries to catch up with his friend
 If anyone stands between friends when one of them moves towards the other, he will receive a slide effect:
 Kama effect - Deal 60 damage and have 25% chanse to cut target leg off,
 Fundo effect - Paralize target on 3 SECONDS.
 Parts can't be picked up if someone else is holding another part.
 All parts can be used in hands to pull it near owner.

## Why It's Good For The Game

Give more funny(and not funny) mechanics for ninja to diversify the gameplay
Energy katana is also useful becouse give teleport. Glaiv also give teleport but with harder mechanics. Kusarigama don't give teleport but give funny gameplay ideas.

:cl:
add: Added 2 new weapons to ninja
/:cl:

Xackii committed:
weaponised

Jackal-boop commented:
Are these alternatives or additions

Xackii commented:
> Are these alternatives or additions

1 of 4. (Writted it in PR)

Xackii committed:
Merge branch 'master' into ningaweapons

Ghostkokos commented:
Based Ninja buffs 👍

cnleth commented:
Why not just allow ninja to use bows when bow is chosen instead of allowing this for everyone with `TRAIT_NOGUNS`?

Xackii commented:
> Why not just allow ninja to use bows when bow is chosen instead of allowing this for everyone with `TRAIT_NOGUNS`?


I can do it this way. It will just be a little strange + the ability to shoot with bows will not change anything much (at most you can shoot with a chaplain’s bow)

Sakamoto4ka commented:
dis bro still cooking

uaioy commented:
something i was told is that no_guns shouldnt let you use any kind of projectile for balance reasons (apparently you can order bows from cargo, which means someone with scarp could use them)

SomeoneYouProbablyKnow commented:
Personally i think the ninja has enough complexity as is. But if you do want to add these different more risky and complex weapons do not present them as an equal option to the katana. The katana should be the path of least resistance for the ninja, you want to funnel newer players towards it, a system where you can fuck yourself over by choosing an unfamiliar weapon is not great.
Make getting those alternative weapons a bit of a process, not a difficult one but just enough of a barrier to make sure the player is making an informed decision.
Maybe make it an easy craft you can do with your katana or perhaps it should be an option you activate from the modsuit panel idk but just something to provide a bit of resistance.

As to the individual weapons
1. It's not that different from a katana but it is a ranged.. knockdown? you didn't specify. But even then coupled with the ninja's stun arms this is scary, potentially too powerful but no way to tell without testing

2. And the bow is not powerful enough, you're trading your teleportation for this. You can def bump up the recharge rate.

Although i am not sure how much we want to allow ranged ninjas with both those weapons

3. A very unique mechanic, but i got a few questions. How does it work with people running after you? Is it the ultimate anti-chase tool, especially considering most people are gonna be confused by the mechanics when they see it in game. Can it be picked up by opponents? 
You do give up your teleport as well but oh well guess i'll just hide the ball under some trash in maits and watch people get stunned for 3 whole seconds walking over it

Loafin34 commented:
two things:
1. this is a feature.
2. the bow feels a little bit underpowered for it's abilities.. maybe give the user a little tanto to finish off enemies after they're stunned by it's repulse effect or something.

Glaive is really cool, but what would make it way cooler is if you could effectively distance yourself by thrusting the glaive at the opponent as an action attack. this could be a way to break up sec congas in maints, where you'll probably be fighting. Also, the bow may be overshadowed a bit by the ranged weapons ninja already has, because shuriken and energy net are both powerful one-on-one dueling devices.

Loafin34 commented on code/modules/antagonists/ninja/weapons/energy_glaive.dm:
```diff
@@ -0,0 +1,110 @@
+#define GLAIVE_COOLDOWN 10 SECONDS
+/obj/item/energy_glaive
+	name = "energy glaive"
+	desc = "Double-sided energy spear."
```
Make the description a little bit cooler. Something like "This glaive is used among guards of the Spider Clan to decimate attacking clans with it's double-bladed design. A small microchip the size of a nail is implanted within the wrist of users of this glaive, which allows the user to invert itself through bluespace to the position of holding the spear. The microchip is activated upon the spear impacting with an object after being thrown.

Loafin34 commented on code/modules/antagonists/ninja/weapons/energy_hankyu.dm:
```diff
@@ -0,0 +1,219 @@
+#define CHARGE_BASE 3 SECONDS
+/obj/item/gun/ballistic/bow/energy_hankyu
+	name = "energy hankyu"
+	desc = "Hi-Tech energy multi-bow"
```
Again, a little lore with it. "Experienced Spider Clan assassins are torn on the fate of the hankyu. Some claim that it's arrows are too bulky and slow to compete with the standard-issue shuriken. however, hankyu users have repeatedly demonstrated the usefulness of the EMP and repulsive arrows against Nanotrasen mooks. Activate in hand to change the arrow payload.

Fikou reviewed (changes requested):
1. This is very unfriendly to new players. Having to handle the janky weight and sickle is not something that will be fun for people picking ninja for their first time especially as nothing is explained to them. You need to add tooltips to the energy case that explain the mechanics of each weapon.
2. The bow is a no-go, as the katana recall module says, "Ranged weapons are forbidden." I don't want ninja to be focused on those.

Fikou commented on code/modules/antagonists/ninja/weapons/energy_glaive.dm:
```diff
@@ -0,0 +1,110 @@
+	inhand_icon_state = "glaive"
+	lefthand_file = 'icons/mob/inhands/weapons/swords_lefthand.dmi'
+	righthand_file = 'icons/mob/inhands/weapons/swords_righthand.dmi'
+	force = 12 // 12 + 12 = 24
```
What is that comment supposed to tell anyone?

Fikou commented on code/modules/antagonists/ninja/weapons/energy_case.dm:
```diff
@@ -0,0 +1,56 @@
+ */
+/obj/item/energy_case
+	name = "energy case"
+	desc = "Black case contains energy weapon"
```
the grammar here doesn't really work

Xackii replied:
what

Fikou commented on code/modules/antagonists/ninja/weapons/energy_case.dm:
```diff
@@ -0,0 +1,56 @@
+	inhand_icon_state = "energy_case"
+	worn_icon_state = "energy_case"
+	w_class = WEIGHT_CLASS_BULKY
+	/// When we choise weapon we auto try to link it with mod weapon recall
+	var/obj/item/link_weapon
```
this doesn't need to be a variable on this object, it can be local to the proc

Fikou replied:
also should be called linked_weapon

Fikou commented on code/modules/antagonists/ninja/weapons/energy_case.dm:
```diff
@@ -0,0 +1,56 @@
+		"Energy Kusarigama" = image(icon = 'icons/obj/weapons/thrown.dmi', icon_state = "energy_kama"),
+		"Energy Hankyu" = image(icon = 'icons/obj/weapons/bows/bows.dmi', icon_state = "hankyu")
+		)
+	var/choise_weapon = show_radial_menu(user, src, weapons, tooltips = TRUE)
```
did you mean chosen? or choice?

Fikou commented on code/modules/antagonists/ninja/weapons/energy_case.dm:
```diff
@@ -0,0 +1,56 @@
+	if(!ishuman(user))
+		return
+	var/mob/living/carbon/human/case_man = user
+	var/obj/item/mod/control/mod = case_man.back
```
its not guaranteed to be on their back at this moment, probably better to do a locate()

Xackii replied:
if it is not on the back, then the mod is not turned on and then there is no point to doing a recall link

Fikou commented on code/modules/antagonists/ninja/weapons/energy_glaive.dm:
```diff
@@ -0,0 +1,110 @@
+	sharpness = SHARP_EDGED
+	max_integrity = 200
+	resistance_flags = LAVA_PROOF | FIRE_PROOF | ACID_PROOF
+	var/datum/effect_system/spark_spread/spark_system
```
autodoc

Fikou commented on code/modules/antagonists/ninja/weapons/energy_glaive.dm:
```diff
@@ -0,0 +1,110 @@
+	max_integrity = 200
+	resistance_flags = LAVA_PROOF | FIRE_PROOF | ACID_PROOF
+	var/datum/effect_system/spark_spread/spark_system
+	/// If if double attacked other target we don't attack first target again
```
```suggestion
	/// Tracks if the next attack will be the automatic double attack to prevent an infinite attacking loop.```

Xackii replied:
if there are no other targets near the attacker, then he attacks the same target again
Priority like:
1. You attacked fist target
2. Try to find second target
3. Attack first target if we haven't found second target

Fikou commented on code/modules/mod/modules/modules_ninja.dm:
```diff
@@ -174,7 +174,13 @@
 	var/obj/item/linked_weapon
 	/// The accepted typepath we can link to.
-	var/accepted_type = /obj/item/energy_katana
+	var/list/accepted_types = list(
```
this module's description is written around the energy katana, you need to rewrite it to make sense

Fikou commented on code/modules/mob/living/carbon/human/human_helpers.dm:
```diff
@@ -118,7 +118,7 @@
 			balloon_alert(src, "fingers are too big!")
 			return FALSE
-	if(HAS_TRAIT(src, TRAIT_NOGUNS))
+	if(HAS_TRAIT(src, TRAIT_NOGUNS) && !istype(G, /obj/item/gun/ballistic/bow)) /// bow is not a gun
```
no guns means no guns, not only some guns

Xackii replied:
Ninja can't use guns

Fikou replied:
yes and i dont want them to

Fikou commented on code/modules/antagonists/ninja/weapons/energy_glaive.dm:
```diff
@@ -0,0 +1,110 @@
+			continue
+		if(another_target == target)
+			continue
+		if(!proximity_flag)
+			continue
```
proximity_flag is not gonna change mid loop. if you want to cancel everything if you click far away then just early return

Fikou commented on code/modules/antagonists/ninja/weapons/energy_glaive.dm:
```diff
@@ -0,0 +1,110 @@
+	charges++
+	balloon_alert(give_ballon, "[charges]/4 ready")
+
+/obj/item/energy_glaive/proc/double_attack(mob/attacker, mob/living/attack_me)
```
user and target will make more sense

Fikou replied:
also there is no guarantee at this point that the user will be holding the glaive, make dropping it cancel the timers

Fikou commented on code/modules/antagonists/ninja/weapons/energy_glaive.dm:
```diff
@@ -0,0 +1,110 @@
+	who.forceMove(where)
+	who.put_in_hands(src)
+	charges--
+	balloon_alert(who, "[charges]/4 left")
```
```suggestion
```
no need to tell them about every charge lost, they will know that from the effect working

Fikou commented on code/modules/antagonists/ninja/weapons/energy_glaive.dm:
```diff
@@ -0,0 +1,110 @@
+
+/obj/item/energy_glaive/proc/add_charge(mob/living/give_ballon)
+	charges++
+	balloon_alert(give_ballon, "[charges]/4 ready")
```
```suggestion
	balloon_alert(give_ballon, "[charges]/4 throw charges")
```

Xackii committed:
Apply suggestions from code review

Co-authored-by: Fikou <23585223+Fikou@users.noreply.github.com>

Xackii commented:
> 1. This is very unfriendly to new players. Having to handle the janky weight and sickle is not something that will be fun for people picking ninja for their first time especially as nothing is explained to them. You need to add tooltips to the energy case that explain the mechanics of each weapon.
> 2. The bow is a no-go, as the katana recall module says, "Ranged weapons are forbidden." I don't want ninja to be focused on those.

1. They're basically just throwing weapons. Although adding more descriptions is still a good idea
2. I didn't understand the problem. The module doesn't work with a bow or you don't like the idea of ​​adding a bow to ninja?

Xackii commented:
> i don't want ninja to have a bow

Very sad. Melbert closed the previous PR on adding traitor bow and suggested adding ninja bow instead. And now another maintainer is telling me that he doesn't want the ninja bow added. In this case, do you want me to add a new weapon or leave 3 to choose from?

Fikou commented:
> do you want me to add a new weapon or leave 3 to choose from?

im not gonna force you to code a fourth weapon, if you think you want to make another one give it a shot

CopperCrow9 commented:
Cool as this is, I see two major problems:
1) The bow is VERY slow. It's more of a utility weapon, but compared to the katana, its a straight downgrade. You don't get the teleport, and you would do more damage with the katana then the arrow anyway.
2) The ball-and-chain weapon would not really work at all in a combat scenario.

I do really like the glaive though, that one looks REALLY cool and I would try that.

Xackii committed:
Merge branch 'master' into ningaweapons

Xackii committed:
everybodylovethebow

Xackii committed:
Merge branch 'ningaweapons' of https://github.com/Xackii/tgstation into ningaweapons

Xackii commented:
removed bow

Xackii committed:
Merge branch 'master' into ningaweapons

Xackii committed:
vadfgaftwatw

Xackii commented:
I corrected all sorts of little things and added more descriptions. Kusarigama can no longer be picked up if one part is held by another person. If a ninja wants to take the kusarigama, then the other part will fall out of the hand of another non-ninja

Xackii committed:
Merge branch 'master' into ningaweapons

Xackii committed:
Merge branch 'master' into ningaweapons

Xackii committed:
Merge branch 'master' into ningaweapons

Xackii committed:
Merge branch 'master' into ningaweapons
